src/Stratis.Bitcoin.Features.Consensus/Rules/CommonRules/PoSTransactionsOrderRule.cs

```diff
@@ -0,0 +1,68 @@
+﻿using Microsoft.Extensions.Logging;
+using NBitcoin;
+using Stratis.Bitcoin.Consensus;
+using Stratis.Bitcoin.Consensus.Rules;
+
+namespace Stratis.Bitcoin.Features.Consensus.Rules.CommonRules
+{
+    /// <summary>Checks that coinstake tx from the proven header is the 2nd transaction in a block and that the block has at least 2 transactions.</summary>
+    public class PosTransactionsOrderRule : IntegrityValidationConsensusRule
+    {
+        private bool isGateway;
+
+        /// <inheritdoc />
+        public override void Initialize()
+        {
+            this.isGateway = this.Parent.ConnectionManagerSettings.IsGateway;
+        }
+
+        /// <inheritdoc />
+        /// <exception cref="ConsensusErrors.InvalidTxCount">Thrown in case block has less than 2 transaction.</exception>
+        /// <exception cref="ConsensusErrors.PHCoinstakeMissmatch">Thrown in case coinstake transaction from the proven header missmatches 2nd block transaction.</exception>
+        /// <exception cref="ConsensusErrors.ProofOfWorkTooHigh">Thrown in case block is a PoW block but created after the last pow height.</exception>
+        public override void Run(RuleContext context)
+        {
+            if (this.isGateway)
+            {
+                this.Logger.LogTrace("(-)[GATEWAY_SKIP]");
+                return;
+            }
+
+            Block block = context.ValidationContext.BlockToValidate;
+
+            var header = context.ValidationContext.ChainedHeaderToValidate.Header as ProvenBlockHeader;
+
+            // In case we are in PoW era there might be no coinstake tx.
+            // We have no way of telling if the block was supposed to be PoW or PoS so attacker
+            // can trick us into thinking that all of them are PoW so no PH is required.
+            if (header == null || !header.Coinstake.IsCoinStake)
```

Comment:
I actually don't understand why the need to fail in case we are bellow the checkpoint.
If this is a regular header this rule does not need to do anything just return.

The purpose of this rule is only to ensure the PH coinstake and bloc coinstake are equal, the other checks don't belong to this rule (and some may even be duplicates)

Reply:
I'm wondering if we should remove this rule when the node is setup as gateway

Reply:
It will be actually more complicated to do that.